<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-03-18" />

<title>Quality control for BAL pool: BP1</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/panelset-0.2.6/panelset.css" rel="stylesheet" />
<script src="site_libs/panelset-0.2.6/panelset.js"></script>
<link href="site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="site_libs/datatables-binding-0.20/datatables.js"></script>
<link href="site_libs/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="site_libs/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="site_libs/dt-core-1.11.3/js/jquery.dataTables.min.js"></script>
<link href="site_libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">cf-eti-bal</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ansontcw/cf-eti-bal">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Quality control for BAL pool: BP1</h1>
<h4 class="author">Anson Wong</h4>
<address class="author_afil">
Molecular Immunity, Murdoch Childrenâ€™s Research
Institute<br><h4 class="date">2025-03-18</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-03-18
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>cf-eti-bal/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R
Markdown file created these results, youâ€™ll want to first commit it to
the Git repo. If youâ€™re still working on the analysis, you can ignore
this warning. When youâ€™re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity itâ€™s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240504code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240504)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240504code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240504)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomansontcwcfetibaltreeadc05cc6441239927c407623f50377379d0e3960targetblankadc05cca">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/ansontcw/cf-eti-bal/tree/adc05cc6441239927c407623f50377379d0e3960" target="_blank">adc05cc</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomansontcwcfetibaltreeadc05cc6441239927c407623f50377379d0e3960targetblankadc05cca"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/ansontcw/cf-eti-bal/tree/adc05cc6441239927c407623f50377379d0e3960" target="_blank">adc05cc</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .pki/
    Ignored:    cf-eti-bal/.DS_Store
    Ignored:    cf-eti-bal/.RData
    Ignored:    cf-eti-bal/analysis/.DS_Store
    Ignored:    cf-eti-bal/analysis/.Rhistory
    Ignored:    cf-eti-bal/analysis/ADGRE_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/ADGRE_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/ANNEXIN_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/ANNEXIN_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/APP_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/APP_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/ApoE_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/ApoE_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/CD45_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/CD45_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/CypA_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/CypA_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/FN1_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/FN1_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/GALECTIN_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/GALECTIN_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/GRN_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/GRN_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/MHC-II_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/MHC-II_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/MHC-I_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/MHC-I_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/MIF_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/MIF_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/MK_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/MK_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/PECAM1_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/PECAM1_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/RESISTIN_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/RESISTIN_hierarchy_individual.png
    Ignored:    cf-eti-bal/analysis/VISFATIN_hierarchy_aggregate.png
    Ignored:    cf-eti-bal/analysis/VISFATIN_hierarchy_individual.png
    Ignored:    cf-eti-bal/data/.DS_Store

Untracked files:
    Untracked:  .Renviron
    Untracked:  .Rprofile
    Untracked:  ._.DS_Store
    Untracked:  ._G000323_Neeland
    Untracked:  .cache/
    Untracked:  .conda/
    Untracked:  .gitattributes
    Untracked:  .gitignore
    Untracked:  .local/
    Untracked:  .subversion/
    Untracked:  C133_Neeland_batch1/
    Untracked:  CF-BAL/
    Untracked:  CF-ETI-scRNA/
    Untracked:  Carraro_2020/
    Untracked:  G000323_Neeland/
    Untracked:  Li_2022/
    Untracked:  Loske_2021/
    Untracked:  SingleCellExperiment_tutorial/
    Untracked:  archive_CF-ETI-scRNA/
    Untracked:  archive_G000323_Neeland/
    Untracked:  archive_cf-eti-bal/
    Untracked:  backup/
    Untracked:  cellranger/
    Untracked:  cf-eti-bal/._.DS_Store
    Untracked:  cf-eti-bal/._RecM.markers.csv
    Untracked:  cf-eti-bal/._Rplots.pdf
    Untracked:  cf-eti-bal/._TRM-CCL.phate.markers.csv
    Untracked:  cf-eti-bal/._TRM-CCL.phate.markers_v250318.csv
    Untracked:  cf-eti-bal/._TRM.markers.csv
    Untracked:  cf-eti-bal/._eti.seurat.metadata.csv
    Untracked:  cf-eti-bal/._pairedRecMProp.csv
    Untracked:  cf-eti-bal/ETI.profibroticScore.fig.jpeg
    Untracked:  cf-eti-bal/RecM.markers.csv
    Untracked:  cf-eti-bal/Rplots.pdf
    Untracked:  cf-eti-bal/TRM-CCL.phate.markers.csv
    Untracked:  cf-eti-bal/TRM-CCL.phate.markers_v250318.csv
    Untracked:  cf-eti-bal/TRM.markers.csv
    Untracked:  cf-eti-bal/analysis/._.DS_Store
    Untracked:  cf-eti-bal/analysis/._04.annotation.Rmd
    Untracked:  cf-eti-bal/analysis/._06.TRM_analysis.Rmd
    Untracked:  cf-eti-bal/analysis/._10.ETI_analysis.Rmd
    Untracked:  cf-eti-bal/analysis/01.demux_BP1.Rmd
    Untracked:  cf-eti-bal/analysis/01.demux_BP2.Rmd
    Untracked:  cf-eti-bal/analysis/01.demux_Pi1.Rmd
    Untracked:  cf-eti-bal/analysis/01.demux_capture1.Rmd
    Untracked:  cf-eti-bal/analysis/02.qc_BP1.Rmd
    Untracked:  cf-eti-bal/analysis/02.qc_BP2.Rmd
    Untracked:  cf-eti-bal/analysis/02.qc_Pi1.Rmd
    Untracked:  cf-eti-bal/analysis/02.qc_capture1.Rmd
    Untracked:  cf-eti-bal/analysis/03.combine.Rmd
    Untracked:  cf-eti-bal/analysis/04.annotation.Rmd
    Untracked:  cf-eti-bal/analysis/05.RecM_analysis.v2.Rmd
    Untracked:  cf-eti-bal/analysis/06.TRM_analysis.Rmd
    Untracked:  cf-eti-bal/analysis/06.TRM_analysis.v2.Rmd
    Untracked:  cf-eti-bal/analysis/07.Epithelial_analysis.Rmd
    Untracked:  cf-eti-bal/analysis/08.TNK_analysis.Rmd
    Untracked:  cf-eti-bal/analysis/09.CD8T_analysis.Rmd
    Untracked:  cf-eti-bal/analysis/09.Others_analysis.Rmd
    Untracked:  cf-eti-bal/analysis/09.cellchat.Rmd
    Untracked:  cf-eti-bal/analysis/09.cellchat.v2.Rmd
    Untracked:  cf-eti-bal/analysis/10.ETI_analysis.Rmd
    Untracked:  cf-eti-bal/analysis/11.MDM-PLA2G7_analysis.Rmd
    Untracked:  cf-eti-bal/analysis/12.TRM_public_analysis.Rmd
    Untracked:  cf-eti-bal/analysis/13.TRM-CCL_analysis.Rmd
    Untracked:  cf-eti-bal/analysis/ADGRE_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/ADGRE_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/ANNEXIN_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/ANNEXIN_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/APP_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/APP_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/ApoE_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/ApoE_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/CD45_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/CD45_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/CypA_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/CypA_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/FN1_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/FN1_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/GALECTIN_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/GALECTIN_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/GRN_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/GRN_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/LAMININ_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/MHC-II_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/MHC-II_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/MHC-I_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/MHC-I_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/MIF_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/MIF_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/MK_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/MK_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/PECAM1_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/PECAM1_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/PHATE_check.Rmd
    Untracked:  cf-eti-bal/analysis/RESISTIN_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/RESISTIN_hierarchy_individual.svg
    Untracked:  cf-eti-bal/analysis/TrialandError.Rmd
    Untracked:  cf-eti-bal/analysis/VISFATIN_hierarchy_aggregate.svg
    Untracked:  cf-eti-bal/analysis/VISFATIN_hierarchy_individual.svg
    Untracked:  cf-eti-bal/before.csv
    Untracked:  cf-eti-bal/capture.list
    Untracked:  cf-eti-bal/clustering.pbs
    Untracked:  cf-eti-bal/code/._bal_RecM.integration_clustering.R
    Untracked:  cf-eti-bal/code/adam.process.R
    Untracked:  cf-eti-bal/code/annotation_HLCAv2.R
    Untracked:  cf-eti-bal/code/azimuth.pbs
    Untracked:  cf-eti-bal/code/bal.manual_annot.integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_CD8T.integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_Epithelial.integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_Epithelial.integration_clustering.round2.R
    Untracked:  cf-eti-bal/code/bal_GOenrichment.R
    Untracked:  cf-eti-bal/code/bal_GOenrichment_noETI.R
    Untracked:  cf-eti-bal/code/bal_GOenrichment_v2.R
    Untracked:  cf-eti-bal/code/bal_GOenrichment_v3.R
    Untracked:  cf-eti-bal/code/bal_GOenrichment_v4.R
    Untracked:  cf-eti-bal/code/bal_Others.integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_RecM.integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_RecM.public_integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_RecM_Control.integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_T.integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_T.integration_clustering.round2.R
    Untracked:  cf-eti-bal/code/bal_TNK.integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_TNK.integration_clustering.round2.R
    Untracked:  cf-eti-bal/code/bal_TNK.integration_clustering.round3.R
    Untracked:  cf-eti-bal/code/bal_TRM-CCL.integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_TRM.integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_TRM.public.CF_integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_TRM.public.HC_integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_TRM.public_integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_TRMimmune.integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_mac.integration_clustering.R
    Untracked:  cf-eti-bal/code/bal_mac.merge_clustering.R
    Untracked:  cf-eti-bal/code/bal_mac.merge_clustering.round2.R
    Untracked:  cf-eti-bal/code/bal_others.merge_clustering.R
    Untracked:  cf-eti-bal/code/bal_tnk.merge_clustering.R
    Untracked:  cf-eti-bal/code/bal_tnk.merge_clustering.round2.R
    Untracked:  cf-eti-bal/code/cellbender.pbs
    Untracked:  cf-eti-bal/code/cellbender.sh
    Untracked:  cf-eti-bal/code/cellranger-multi.sh
    Untracked:  cf-eti-bal/code/cellsnp-lite.pbs
    Untracked:  cf-eti-bal/code/cellsnp-lite.sh
    Untracked:  cf-eti-bal/code/cellsnp-lite_emptyDrops.pbs
    Untracked:  cf-eti-bal/code/cellsnp-lite_emptyDrops.sh
    Untracked:  cf-eti-bal/code/clustering.R
    Untracked:  cf-eti-bal/code/clustering_try.R
    Untracked:  cf-eti-bal/code/integration.R
    Untracked:  cf-eti-bal/code/integration_batch.R
    Untracked:  cf-eti-bal/code/integration_batchID.R
    Untracked:  cf-eti-bal/code/integration_clustering.R
    Untracked:  cf-eti-bal/code/mac.subcluster.R
    Untracked:  cf-eti-bal/code/qsub_azimuth.sh
    Untracked:  cf-eti-bal/code/qsub_cellbender.sh
    Untracked:  cf-eti-bal/code/qsub_cellsnp-lite.sh
    Untracked:  cf-eti-bal/code/qsub_cellsnp-lite_emptyDrops.sh
    Untracked:  cf-eti-bal/code/qsub_vireo.sh
    Untracked:  cf-eti-bal/code/qsub_vireo_emptyDrops.sh
    Untracked:  cf-eti-bal/code/vireo.pbs
    Untracked:  cf-eti-bal/code/vireo.sh
    Untracked:  cf-eti-bal/code/vireo_emptyDrops.pbs
    Untracked:  cf-eti-bal/code/vireo_emptyDrops.sh
    Untracked:  cf-eti-bal/code/writeDE.R
    Untracked:  cf-eti-bal/data.use.csv
    Untracked:  cf-eti-bal/data/._.DS_Store
    Untracked:  cf-eti-bal/data/._Fig2B.raw.csv
    Untracked:  cf-eti-bal/data/._Fig2B.stat.CFvsHC.csv
    Untracked:  cf-eti-bal/data/._Fig2B.stat.csv
    Untracked:  cf-eti-bal/data/._Fig3A.raw.csv
    Untracked:  cf-eti-bal/data/._Fig3A.stat.csv
    Untracked:  cf-eti-bal/data/._Li.stat.csv
    Untracked:  cf-eti-bal/data/._Liao.stat.csv
    Untracked:  cf-eti-bal/data/._Morrell.stat.csv
    Untracked:  cf-eti-bal/data/._TRM.phate.Cepo.markers.csv
    Untracked:  cf-eti-bal/data/._TableE2.csv
    Untracked:  cf-eti-bal/data/._Wendisch.stat.csv
    Untracked:  cf-eti-bal/data/._cftr.exp.csv
    Untracked:  cf-eti-bal/data/CellChat/
    Untracked:  cf-eti-bal/data/DE/
    Untracked:  cf-eti-bal/data/Fig2B.raw.csv
    Untracked:  cf-eti-bal/data/Fig2B.stat.CFvsHC.csv
    Untracked:  cf-eti-bal/data/Fig2B.stat.csv
    Untracked:  cf-eti-bal/data/Fig3A.raw.csv
    Untracked:  cf-eti-bal/data/Fig3A.stat.csv
    Untracked:  cf-eti-bal/data/FindAllMarkers/
    Untracked:  cf-eti-bal/data/GOBP/
    Untracked:  cf-eti-bal/data/Li.stat.csv
    Untracked:  cf-eti-bal/data/Liao.stat.csv
    Untracked:  cf-eti-bal/data/Morrell.stat.csv
    Untracked:  cf-eti-bal/data/SCEs/
    Untracked:  cf-eti-bal/data/TRM.phate.Cepo.markers.csv
    Untracked:  cf-eti-bal/data/TableE2.csv
    Untracked:  cf-eti-bal/data/Wendisch.stat.csv
    Untracked:  cf-eti-bal/data/cellbender/
    Untracked:  cf-eti-bal/data/cellranger/
    Untracked:  cf-eti-bal/data/cellsnp-lite/
    Untracked:  cf-eti-bal/data/cellsnp-lite_emptyDrops/
    Untracked:  cf-eti-bal/data/cftr.exp.csv
    Untracked:  cf-eti-bal/data/emptyDrops/
    Untracked:  cf-eti-bal/data/pathway/
    Untracked:  cf-eti-bal/data/plots/
    Untracked:  cf-eti-bal/data/public_datasets/
    Untracked:  cf-eti-bal/data/sample_sheets/
    Untracked:  cf-eti-bal/data/tradeSeq/
    Untracked:  cf-eti-bal/data/tradeseq/
    Untracked:  cf-eti-bal/data/vireo/
    Untracked:  cf-eti-bal/data/vireo_emptyDrops/
    Untracked:  cf-eti-bal/eti.seurat.metadata.csv
    Untracked:  cf-eti-bal/integration.pbs
    Untracked:  cf-eti-bal/minmax.csv
    Untracked:  cf-eti-bal/notes/
    Untracked:  cf-eti-bal/obsolete/
    Untracked:  cf-eti-bal/pairedRecMProp.csv
    Untracked:  cf-eti-bal/pairedRecMProp.jpeg
    Untracked:  cf-eti-bal/phate_batch.jpeg
    Untracked:  cf-eti-bal/phate_dens_batch.jpeg
    Untracked:  cf-eti-bal/phate_dens_condition.jpeg
    Untracked:  cf-eti-bal/phate_fastMNN.jpeg
    Untracked:  cf-eti-bal/qsub_clustering.sh
    Untracked:  cf-eti-bal/qsub_integration.sh
    Untracked:  cf-eti-bal/queue_log/
    Untracked:  cf-eti-bal/vireo.list
    Untracked:  cf-eti-bal/workflowr.sh
    Untracked:  cf-eti-bal/zscore.csv
    Untracked:  cf-pbmc-bal/
    Untracked:  chapter3/
    Untracked:  chapter4/
    Untracked:  healthy-bal/
    Untracked:  presentation/

Staged changes:
    New:        cf-eti-bal/.Rprofile
    New:        cf-eti-bal/.gitattributes
    New:        cf-eti-bal/.gitignore
    New:        cf-eti-bal/README.md
    New:        cf-eti-bal/_workflowr.yml
    New:        cf-eti-bal/analysis/about.Rmd
    New:        cf-eti-bal/analysis/index.Rmd
    New:        cf-eti-bal/analysis/license.Rmd
    New:        cf-eti-bal/cf-eti-bal.Rproj
    New:        cf-eti-bal/code/README.md
    New:        cf-eti-bal/data/README.md
    New:        cf-eti-bal/output/README.md

</code></pre>
<p>
Note that any generated files, e.g.Â HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<pre class="r"><code>knitr::opts_chunk$set(warning = FALSE, message = FALSE)
xaringanExtra::use_panelset()</code></pre>
<div id="load-packages" class="section level1">
<h1>1 Load packages</h1>
<pre class="r"><code>suppressPackageStartupMessages({
library(DropletUtils)
library(here)
library(ggplot2)
library(Seurat)
library(cowplot)
library(patchwork)
library(scater)
library(dplyr)
library(forcats)
library(janitor)
library(stringr)
library(AnnotationHub)
library(ensembldb)
library(msigdbr)
library(Homo.sapiens)
})
set.seed(1990)</code></pre>
</div>
<div id="set-names-for-this-pool" class="section level1">
<h1>2 Set names for this pool</h1>
<pre class="r"><code># Specify batch name
batch_name &lt;- &quot;BP1&quot;
num_of_captures &lt;- 3

# Specify capture name
capture_names &lt;- c(paste0(batch_name, &quot;-c&quot;,1:num_of_captures))
capture_names &lt;- setNames(capture_names, capture_names)

# Assign sample ID to HTO ID
# Manually list sample names matching HTO id (in alphabetical order)
samples &lt;- c(&quot;M1C198B&quot;, #HTO3
             &quot;M1C208A&quot;, #HTO6
             &quot;M1C166(1)&quot;, #HTO7
             &quot;M1C199B&quot;, #HTO8
             &quot;M1C207B&quot;, #HTO10
             &quot;M1N087&quot;, #HTO12
             &quot;M1N080&quot; #HTO13
             )</code></pre>
</div>
<div id="filter-out-ambiguous-droplets" class="section level1">
<h1>3 Filter out ambiguous droplets</h1>
<pre class="r"><code># read demultiplexing outputs
sce &lt;- readRDS(here(&quot;data&quot;,
                    &quot;SCEs&quot;,
                    &quot;demux&quot;,
                    paste0(batch_name,&quot;.cellbender.demux.SCE.rds&quot;)))

# number of droplets before removal
dim(sce)</code></pre>
<pre><code>[1] 36601 62050</code></pre>
<pre class="r"><code># A. remove droplets classified as &quot;doublets&quot; by Vireo
sce &lt;- sce[, which(sce$sampleID.genetics != &quot;Doublet&quot;)]

# number of droplets after doublet removal
dim(sce)</code></pre>
<pre><code>[1] 36601 55465</code></pre>
<pre class="r"><code># look for the following droplets
## genetically non-negatives with matched sampleID.genetics and sampleID.HTO
## or classified as Doublet or Negative by HTODemux
nneg_match &lt;- sce$sampleID.genetics != &quot;Negative&quot; &amp; 
  (sce$sampleID.HTO == sce$sampleID.genetics | 
     sce$sampleID.HTO %in% c(&quot;Doublet&quot;,&quot;Negative&quot;))

## genetically negatives but with HTO classification
neg_match &lt;- sce$sampleID.genetics == &quot;Negative&quot; &amp; 
  sce$sampleID.HTO %in% samples

# B. keep droplets that match the demultiplexing criteria
sce &lt;- sce[, nneg_match | neg_match]

# number of droplets after demultiplexing
dim(sce)</code></pre>
<pre><code>[1] 36601 49781</code></pre>
<pre class="r"><code># C. remove droplets with posterior counts of zero called by CellBender
# see: https://github.com/broadinstitute/CellBender/issues/111
sce &lt;- sce[, colSums(counts(sce)) != 0]

# number of droplets after removing droplets with zero counts
dim(sce)</code></pre>
<pre><code>[1] 36601 49210</code></pre>
<pre class="r"><code># rename sampleID
sampleID &lt;- setNames(factor(case_when(
  sce$sampleID.genetics == &quot;Negative&quot; ~ sce$sampleID.HTO,
  TRUE ~ sce$sampleID.genetics),levels = samples),
  colnames(sce))

sce$sampleID &lt;- sampleID</code></pre>
</div>
<div id="visualize-results-after-filtering" class="section level1">
<h1>4 Visualize results after filtering</h1>
<pre class="r"><code># number of droplets assigned by HTO method
p1 &lt;- ggcells(sce) + 
  geom_bar(aes(x = sampleID.HTO, fill = sampleID.HTO)) + 
  geom_text(stat=&#39;count&#39;, aes(x = sampleID.HTO, label=..count..), hjust=1, size=2) +
  coord_flip() + 
  ggtitle(&quot;By HTO&quot;) +
  ylab(&quot;Number of droplets&quot;) + 
  theme_cowplot(font_size = 8) + 
  scale_y_continuous(breaks=seq(0,12000,3000),limits=c(0,16000)) +
  scale_fill_manual(values = sce$colours$sampleID.HTO_colours) +
  theme(axis.title.y = element_blank()) +
  guides(fill=guide_legend(title=&quot;sampleID&quot;))

p1.facet &lt;- ggcells(sce) + 
  geom_bar(aes(x = sampleID.HTO, fill = sampleID.HTO)) + 
  geom_text(stat=&#39;count&#39;, aes(x = sampleID.HTO, label=..count..), 
            hjust=1, size=2) +
  #ggtitle(&quot;By HTO&quot;) +
  ylab(&quot;Number of droplets&quot;) + 
  facet_grid(~Capture, scales = &quot;fixed&quot;, space = &quot;fixed&quot;) + 
  theme_cowplot(font_size = 8) + 
  scale_y_continuous(breaks=seq(0,6000,2000), limits = c(0,6000)) +
  scale_fill_manual(values = sce$colours$sampleID.HTO_colours) +
  theme(axis.title.y = element_blank()) +
  guides(fill=guide_legend(title=&quot;sampleID&quot;)) +
    coord_flip()

# number of droplets assigned by genetic method
p2 &lt;- ggcells(sce) + 
  geom_bar(aes(x = sampleID.genetics, fill = sampleID.genetics)) + 
  geom_text(stat=&#39;count&#39;, aes(x = sampleID.genetics, label=..count..), hjust=1, size=2) +
  coord_flip() + 
  ggtitle(&quot;By genetics&quot;) +
  ylab(&quot;Number of droplets&quot;) + 
  theme_cowplot(font_size = 8) + 
  scale_y_continuous(breaks=seq(0,12000,3000), limits = c(0,16000)) +
  scale_fill_manual(values = sce$colours$sampleID.genetics_colours) +
  theme(axis.title.y = element_blank()) +
  guides(fill=FALSE)

p2.facet &lt;- ggcells(sce) + 
  geom_bar(aes(x = sampleID.genetics, fill = sampleID.genetics)) + 
  geom_text(stat=&#39;count&#39;, aes(x = sampleID.genetics, label=..count..), 
            hjust=1, size=2) +
  ylab(&quot;Number of droplets&quot;) + 
  facet_grid(~Capture, scales = &quot;fixed&quot;, space = &quot;fixed&quot;) + 
  theme_cowplot(font_size = 8) + 
  scale_y_continuous(breaks=seq(0,6000,2000), limits = c(0,6000)) +
  scale_fill_manual(values = sce$colours$sampleID.genetics_colours) +
  theme(axis.title.y = element_blank()) +
  guides(fill=guide_legend(title=&quot;sampleID&quot;)) +
  coord_flip()

(p1+p1.facet+plot_layout(width=c(1,2))) /
  (p2+p2.facet+plot_layout(width=c(1,2))) +
  plot_layout(guides=&quot;collect&quot;)</code></pre>
<p><img src="figure/02.qc_BP1.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># proportion of genetically assigned droplets in each HTO
p3 &lt;- ggcells(sce) + 
  geom_bar(
    aes(x = sampleID.HTO, fill = sampleID.genetics),
    position = position_fill(reverse = TRUE)) +
  coord_flip() +
  ylab(&quot;Frequency&quot;) +
  theme_cowplot(font_size = 8) + 
  scale_fill_manual(values = sce$colours$sampleID.genetics_colours)

# proportion of HTO assigned droplets in each genetic donor
p4 &lt;- ggcells(sce) + 
  geom_bar(
    aes(x = sampleID.genetics, fill = sampleID.HTO),
    position = position_fill(reverse = TRUE)) +
  coord_flip() +
  ylab(&quot;Frequency&quot;) +
  theme_cowplot(font_size = 8) + 
  scale_fill_manual(values = sce$colours$sampleID.HTO_colours)

((p3 + ggtitle(&quot;By HTO&quot;)) +
    p3 + facet_grid(~Capture) + plot_layout(widths = c(1, 2))) /
  ((p4 + ggtitle(&quot;By genetics&quot;)) + 
     p4 + facet_grid(~Capture) + plot_layout(widths = c(1, 2))) +
  plot_layout(guides = &quot;collect&quot;)</code></pre>
<p><img src="figure/02.qc_BP1.Rmd/unnamed-chunk-4-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="add-metadata-and-sex-check" class="section level1">
<h1>5 Add metadata and sex check</h1>
<pre class="r"><code># read metadata.csv
metadata &lt;- read.csv(here(&quot;data&quot;,
                          &quot;sample_sheets&quot;,
                          paste0(batch_name,&quot;.metadata.csv&quot;)))

i &lt;- match(sce$sampleID, metadata$sampleID)

# add patient demographics
colData(sce) &lt;- cbind(
  colData(sce),
  metadata[i,c(&quot;Age&quot;,&quot;Sex&quot;,&quot;Condition&quot;,&quot;Bronchiectasis&quot;)]
)

# sex check by the expression of XIST, a female-specific gene. 
# this detects certain types of sample-mix-ups.
plotExpression(
  sce, 
  &quot;XIST&quot;, 
  x = &quot;sampleID&quot;,
  colour_by = &quot;Sex&quot;, 
  exprs_values = &quot;counts&quot;,
  swap_rownames = &quot;Symbol&quot;) +
  scale_colour_manual(
    values = c(
      &quot;Female&quot; = &quot;deeppink3&quot;, 
      &quot;Male&quot; = &quot;deepskyblue&quot;),
    name = &quot;Sex&quot;) +
  theme_cowplot() + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))</code></pre>
<p><img src="figure/02.qc_BP1.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="add-feature-based-annotation" class="section level1">
<h1>6 Add feature-based annotation</h1>
<pre class="r"><code># make feature names unique
rownames(sce) &lt;- uniquifyFeatureNames(rowData(sce)$ID, rowData(sce)$Symbol)

# prepare ensembl v98 database
ah &lt;- AnnotationHub(cache=&quot;/group/canc2/anson/.cache/R/AnnotationHub&quot;, ask=FALSE)
EnsDb.Hsapiens.v98 &lt;- query(ah, c(&quot;EnsDb&quot;, &quot;Homo Sapiens&quot;, 98))[[1]]

ensdb_columns &lt;- setNames(c(&quot;GENEBIOTYPE&quot;, &quot;SEQNAME&quot;),
                          paste0(&quot;ENSEMBL.&quot;,
                                 c(&quot;GENEBIOTYPE&quot;, &quot;SEQNAME&quot;)))
stopifnot(all(ensdb_columns %in% columns(EnsDb.Hsapiens.v98)))

ensdb_df &lt;- DataFrame(
  lapply(ensdb_columns, function(column) {
    mapIds(
      x = EnsDb.Hsapiens.v98,
      keys = rowData(sce)$ID,
      keytype = &quot;GENEID&quot;,
      column = column,
      multiVals = &quot;first&quot;)
  }),
  row.names = rowData(sce)$ID)

# prepare ncbi database
ncbi_columns &lt;- setNames(c(&quot;ALIAS&quot;, &quot;ENTREZID&quot;, &quot;GENENAME&quot;),
                         paste0(&quot;NCBI.&quot;, c(&quot;ALIAS&quot;, &quot;ENTREZID&quot;, &quot;GENENAME&quot;)))
stopifnot(all(ncbi_columns %in% columns(Homo.sapiens)))
ncbi_df &lt;- DataFrame(
  lapply(ncbi_columns, function(column) {
    mapIds(
      x = Homo.sapiens,
      keys = rowData(sce)$ID,
      keytype = &quot;ENSEMBL&quot;,
      column = column,
      multiVals = &quot;CharacterList&quot;)
  }),
  row.names = rowData(sce)$ID)

rowData(sce) &lt;- cbind(rowData(sce), ensdb_df, ncbi_df)</code></pre>
</div>
<div id="qc" class="section level1">
<h1>7 QC</h1>
<p>Some code in this section are derived from Dr.Â Jovana Maksimovicâ€™s
work for (Maksimovic et al.Â 2022)</p>
<div id="calculate-metrics" class="section level2">
<h2>7.1 Calculate metrics</h2>
<pre class="r"><code># prepare gene sets
## mitochondrial gene set
mito_set &lt;- rownames(sce)[which(rowData(sce)$ENSEMBL.SEQNAME == &quot;MT&quot;)]
is_mito &lt;- rownames(sce) %in% mito_set
summary(is_mito)</code></pre>
<pre><code>   Mode   FALSE    TRUE 
logical   36588      13 </code></pre>
<pre class="r"><code>## ribosomal gene set
ribo_set &lt;- grep(&quot;^RP(S|L)&quot;, rownames(sce), value = TRUE)
c2_sets &lt;- msigdbr(species = &quot;Homo sapiens&quot;, category = &quot;C2&quot;)
ribo_set &lt;- union(
  ribo_set,
  c2_sets[c2_sets$gs_name == &quot;KEGG_RIBOSOME&quot;, ]$human_gene_symbol)
is_ribo &lt;- rownames(sce) %in% ribo_set
summary(is_ribo)</code></pre>
<pre><code>   Mode   FALSE    TRUE 
logical   36494     107 </code></pre>
<pre class="r"><code>## sex-linked genes
sex_set &lt;- rownames(sce)[rowData(sce)$ENSEMBL.SEQNAME %in% c(&quot;X&quot;, &quot;Y&quot;)]

## pseudogenes
pseudogene_set &lt;- rownames(sce)[grepl(&quot;pseudogene&quot;, rowData(sce)$ENSEMBL.GENEBIOTYPE)]

# calculate QC metrics
sce &lt;- addPerCellQCMetrics(
  sce,
  subsets = list(Mito = which(is_mito),
                 Ribo = which(is_ribo)), 
  use.altexps = NULL)</code></pre>
</div>
<div id="qc-ncount-and-nfeature" class="section level2">
<h2>7.2 QC: nCount and nFeature</h2>
<div id="visualize-metrics-before-removal" class="section level3">
<h3>7.2.1 Visualize metrics before removal</h3>
<pre class="r"><code># library size
p1 &lt;- plotColData(
  sce,
  &quot;sum&quot;,
  x = &quot;sampleID&quot;,
  colour_by = &quot;sampleID&quot;,
  point_size = 0.5) +
  scale_y_log10() +
  scale_colour_manual(values = sce$colours$sample_colours, name = &quot;sampleID&quot;) +
  theme(axis.text.x = element_blank())

# number of genes detected
p2 &lt;- plotColData(
  sce,
  &quot;detected&quot;,
  x = &quot;sampleID&quot;,
  colour_by = &quot;sampleID&quot;,
  point_size = 0.5) +
  scale_colour_manual(values = sce$colours$sample_colours, name = &quot;sampleID&quot;) +
  theme(axis.text.x = element_blank())

((p1 + NoLegend()) + p1 + facet_grid(~sce$Capture) + plot_layout(widths=c(1, length(capture_names))))/
((p2 + NoLegend()) + p2 + facet_grid(~sce$Capture) + plot_layout(widths=c(1, length(capture_names)))) + 
  plot_layout(guides=&quot;collect&quot;)</code></pre>
<p><img src="figure/02.qc_BP1.Rmd/unnamed-chunk-8-1.png" width="960" style="display: block; margin: auto;" />
### 7.2.2 Remove cells with features &lt; 200</p>
<pre class="r"><code>sce$batch &lt;- interaction(
  sce$Capture, 
  sce$sampleID, 
  drop = TRUE,
  lex.order = FALSE)

feature_drop &lt;- sce$detected &lt; 200

sce_pre_QC_outlier_removal &lt;- sce
keep &lt;- !feature_drop
sce_pre_QC_outlier_removal$keep &lt;- keep
sce &lt;- sce[, keep]

data.frame(
  ByFeature = tapply(
    feature_drop, 
    sce_pre_QC_outlier_removal$batch, 
    sum,
    na.rm = TRUE),
  Remaining = as.vector(unname(table(sce$batch))),
  PercRemaining = round(
    100 * as.vector(unname(table(sce$batch))) /
      as.vector(
        unname(
          table(sce_pre_QC_outlier_removal$batch))), 1)) |&gt;
  tibble::rownames_to_column(&quot;batch&quot;) |&gt;
  dplyr::arrange(dplyr::desc(PercRemaining)) |&gt;
  DT::datatable(
    caption = &quot;Number of droplets removed by each QC step and the number of droplets remaining.&quot;,
    rownames = FALSE) |&gt;
  DT::formatRound(&quot;PercRemaining&quot;, 1)</code></pre>
<div class="datatables html-widget html-fill-item" id="htmlwidget-3046410321596ae3a260" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-3046410321596ae3a260">{"x":{"filter":"none","vertical":false,"caption":"<caption>Number of droplets removed by each QC step and the number of droplets remaining.<\/caption>","data":[["BP1-c3.M1N080","BP1-c1.M1C198B","BP1-c3.M1C198B","BP1-c2.M1N080","BP1-c1.M1N087","BP1-c1.M1N080","BP1-c2.M1C198B","BP1-c3.M1N087","BP1-c1.M1C208A","BP1-c2.M1N087","BP1-c3.M1C208A","BP1-c2.M1C208A","BP1-c1.M1C166(1)","BP1-c3.M1C166(1)","BP1-c2.M1C166(1)","BP1-c1.M1C207B","BP1-c3.M1C207B","BP1-c3.M1C199B","BP1-c2.M1C207B","BP1-c1.M1C199B","BP1-c2.M1C199B"],[103,203,203,137,212,195,226,215,458,216,452,516,404,362,379,454,534,263,512,315,271],[1793,2990,2529,1673,2342,2130,2310,2129,4196,1803,3618,3355,1841,1628,1569,1631,1550,737,1382,771,603],[94.59999999999999,93.59999999999999,92.59999999999999,92.40000000000001,91.7,91.59999999999999,91.09999999999999,90.8,90.2,89.3,88.90000000000001,86.7,82,81.8,80.5,78.2,74.40000000000001,73.7,73,71,69]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>batch<\/th>\n      <th>ByFeature<\/th>\n      <th>Remaining<\/th>\n      <th>PercRemaining<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":3,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 1, 3, \",\", \".\");\n  }"},{"className":"dt-right","targets":[1,2,3]}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render"],"jsHooks":[]}</script>
</div>
<div id="visualize-metrics-after-removal" class="section level3">
<h3>7.2.3 Visualize metrics after removal</h3>
<pre class="r"><code>p3 &lt;- plotColData(
  sce_pre_QC_outlier_removal,
  &quot;sum&quot;,
  x = &quot;sampleID&quot;,
  colour_by = &quot;keep&quot;,
  point_size = 0.5) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
p4 &lt;- plotColData(
  sce_pre_QC_outlier_removal,
  &quot;detected&quot;,
  x = &quot;sampleID&quot;,
  colour_by = &quot;keep&quot;,
  point_size = 0.5) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

((p3 + NoLegend()) + p3 + facet_grid(~sce_pre_QC_outlier_removal$Capture) + plot_layout(widths=c(1, length(capture_names))))/
((p4 + NoLegend()) + p4 + facet_grid(~sce_pre_QC_outlier_removal$Capture) + plot_layout(widths=c(1, length(capture_names)))) + 
  plot_layout(guides=&quot;collect&quot;)</code></pre>
<p><img src="figure/02.qc_BP1.Rmd/unnamed-chunk-10-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="qc-mitochondrial-percent" class="section level2">
<h2>7.3 QC: mitochondrial percent</h2>
<div id="visualize-metrics-before-removal-1" class="section level3">
<h3>7.3.1 Visualize metrics before removal</h3>
<pre class="r"><code># subsets_Mito_percent
p1 &lt;- plotColData(
  sce,
  &quot;subsets_Mito_percent&quot;,
  x = &quot;sampleID&quot;,
  colour_by = &quot;sampleID&quot;,
  point_size = 0.5) +
  scale_y_log10() +
  scale_colour_manual(values = sce$colours$sample_colours, name = &quot;sampleID&quot;) +
  theme(axis.text.x = element_blank())

# subsets_Ribo_percent
p2 &lt;- plotColData(
  sce,
  &quot;subsets_Ribo_percent&quot;,
  x = &quot;sampleID&quot;,
  colour_by = &quot;sampleID&quot;,
  point_size = 0.5) +
  scale_colour_manual(values = sce$colours$sample_colours, name = &quot;sampleID&quot;) +
  theme(axis.text.x = element_blank())

((p1 + NoLegend()) + p1 + facet_grid(~sce$Capture) + plot_layout(widths=c(1, length(capture_names))))/
((p2 + NoLegend()) + p2 + facet_grid(~sce$Capture) + plot_layout(widths=c(1, length(capture_names)))) + 
  plot_layout(guides=&quot;collect&quot;)</code></pre>
<p><img src="figure/02.qc_BP1.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="remove-mito-outlier" class="section level3">
<h3>7.3.2 Remove mito% outlier</h3>
<pre class="r"><code>mito_drop &lt;- isOutlier(
  metric = sce$subsets_Mito_percent, 
  nmads = 3, 
  type = &quot;higher&quot;,
  batch = sce$batch)

sce_pre_QC_outlier_removal &lt;- sce
keep &lt;- !mito_drop
sce_pre_QC_outlier_removal$keep &lt;- keep
sce &lt;- sce[, keep]

data.frame(
  ByMito = tapply(
    mito_drop, 
    sce_pre_QC_outlier_removal$batch, 
    sum,
    na.rm = TRUE),
  Remaining = as.vector(unname(table(sce$batch))),
  PercRemaining = round(
    100 * as.vector(unname(table(sce$batch))) /
      as.vector(
        unname(
          table(sce_pre_QC_outlier_removal$batch))), 1)) |&gt;
  tibble::rownames_to_column(&quot;batch&quot;) |&gt;
  dplyr::arrange(dplyr::desc(PercRemaining)) |&gt;
  DT::datatable(
    caption = &quot;Number of droplets removed by each QC step and the number of droplets remaining.&quot;,
    rownames = FALSE) |&gt;
  DT::formatRound(&quot;PercRemaining&quot;, 1)</code></pre>
<div class="datatables html-widget html-fill-item" id="htmlwidget-07331b92d116fa557da5" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-07331b92d116fa557da5">{"x":{"filter":"none","vertical":false,"caption":"<caption>Number of droplets removed by each QC step and the number of droplets remaining.<\/caption>","data":[["BP1-c2.M1N080","BP1-c3.M1N080","BP1-c2.M1C198B","BP1-c3.M1N087","BP1-c1.M1C198B","BP1-c1.M1N080","BP1-c1.M1N087","BP1-c3.M1C198B","BP1-c3.M1C208A","BP1-c2.M1C208A","BP1-c2.M1N087","BP1-c1.M1C208A","BP1-c1.M1C199B","BP1-c3.M1C166(1)","BP1-c1.M1C166(1)","BP1-c2.M1C166(1)","BP1-c2.M1C207B","BP1-c2.M1C199B","BP1-c3.M1C207B","BP1-c3.M1C199B","BP1-c1.M1C207B"],[67,75,99,94,134,96,113,123,193,182,98,243,52,125,146,129,116,52,133,67,160],[1606,1718,2211,2035,2856,2034,2229,2406,3425,3173,1705,3953,719,1503,1695,1440,1266,551,1417,670,1471],[96,95.8,95.7,95.59999999999999,95.5,95.5,95.2,95.09999999999999,94.7,94.59999999999999,94.59999999999999,94.2,93.3,92.3,92.09999999999999,91.8,91.59999999999999,91.40000000000001,91.40000000000001,90.90000000000001,90.2]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>batch<\/th>\n      <th>ByMito<\/th>\n      <th>Remaining<\/th>\n      <th>PercRemaining<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":3,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 1, 3, \",\", \".\");\n  }"},{"className":"dt-right","targets":[1,2,3]}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render"],"jsHooks":[]}</script>
</div>
</div>
<div id="visualize-metrics-after-removal-1" class="section level2">
<h2>7.3.1 Visualize metrics after removal</h2>
<pre class="r"><code>p3 &lt;- plotColData(
  sce_pre_QC_outlier_removal,
  &quot;subsets_Mito_percent&quot;,
  x = &quot;sampleID&quot;,
  colour_by = &quot;keep&quot;,
  point_size = 0.5) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

p4 &lt;- plotColData(
  sce_pre_QC_outlier_removal,
  &quot;subsets_Ribo_percent&quot;,
  x = &quot;sampleID&quot;,
  colour_by = &quot;keep&quot;,
  point_size = 0.5) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

((p3 + NoLegend()) + p3 + facet_grid(~sce_pre_QC_outlier_removal$Capture) + plot_layout(widths=c(1, length(capture_names))))/
((p4 + NoLegend()) + p4 + facet_grid(~sce_pre_QC_outlier_removal$Capture) + plot_layout(widths=c(1, length(capture_names)))) + 
  plot_layout(guides=&quot;collect&quot;)</code></pre>
<p><img src="figure/02.qc_BP1.Rmd/unnamed-chunk-13-1.png" width="960" style="display: block; margin: auto;" /></p>
<div id="number-of-droplets-retained-named-by-sample-id"
class="section level3">
<h3>7.5.2 Number of droplets retained (named by sample ID)</h3>
<pre class="r"><code># number of droplets assigned by HTO method
p1 &lt;- ggcells(sce) + 
  geom_bar(aes(x = sampleID.HTO, fill = sampleID.HTO)) + 
  geom_text(stat=&#39;count&#39;, aes(x = sampleID.HTO, label=..count..), hjust=1, size=2) +
  coord_flip() + 
  ggtitle(&quot;By HTO&quot;) +
  ylab(&quot;Number of droplets&quot;) + 
  theme_cowplot(font_size = 8) + 
  scale_y_continuous(breaks=seq(0,16000,4000),limits=c(0,16000)) +
  scale_fill_manual(values = sce$colours$sampleID.HTO_colours) +
  theme(axis.title.y = element_blank()) +
  guides(fill=guide_legend(title=&quot;sampleID&quot;))

p1.facet &lt;- ggcells(sce) + 
  geom_bar(aes(x = sampleID.HTO, fill = sampleID.HTO)) + 
  geom_text(stat=&#39;count&#39;, aes(x = sampleID.HTO, label=..count..), 
            hjust=1, size=2) +
  #ggtitle(&quot;By HTO&quot;) +
  ylab(&quot;Number of droplets&quot;) + 
  facet_grid(~Capture, scales = &quot;fixed&quot;, space = &quot;fixed&quot;) + 
  theme_cowplot(font_size = 8) + 
  scale_y_continuous(breaks=seq(0,4000,1000), limits = c(0,4000)) +
  scale_fill_manual(values = sce$colours$sampleID.HTO_colours) +
  theme(axis.title.y = element_blank()) +
  guides(fill=guide_legend(title=&quot;sampleID&quot;)) +
    coord_flip()

# number of droplets assigned by genetic method
p2 &lt;- ggcells(sce) + 
  geom_bar(aes(x = sampleID.genetics, fill = sampleID.genetics)) + 
  geom_text(stat=&#39;count&#39;, aes(x = sampleID.genetics, label=..count..), hjust=1, size=2) +
  coord_flip() + 
  ggtitle(&quot;By genetics&quot;) +
  ylab(&quot;Number of droplets&quot;) + 
  theme_cowplot(font_size = 8) + 
  scale_y_continuous(breaks=seq(0,16000,4000), limits = c(0,16000)) +
  scale_fill_manual(values = sce$colours$sampleID.genetics_colours) +
  theme(axis.title.y = element_blank()) +
  guides(fill=FALSE)

p2.facet &lt;- ggcells(sce) + 
  geom_bar(aes(x = sampleID.genetics, fill = sampleID.genetics)) + 
  geom_text(stat=&#39;count&#39;, aes(x = sampleID.genetics, label=..count..), 
            hjust=1, size=2) +
  ylab(&quot;Number of droplets&quot;) + 
  facet_grid(~Capture, scales = &quot;fixed&quot;, space = &quot;fixed&quot;) + 
  theme_cowplot(font_size = 8) + 
  scale_y_continuous(breaks=seq(0,4000,1000), limits = c(0,4000)) +
  scale_fill_manual(values = sce$colours$sampleID.genetics_colours) +
  theme(axis.title.y = element_blank()) +
  guides(fill=guide_legend(title=&quot;sampleID&quot;)) +
  coord_flip()

(p1+p1.facet+plot_layout(width=c(1,2))) /
  (p2+p2.facet+plot_layout(width=c(1,2))) +
  plot_layout(guides=&quot;collect&quot;)</code></pre>
<p><img src="figure/02.qc_BP1.Rmd/unnamed-chunk-14-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="proportion-of-droplets-named-by-sample-id"
class="section level3">
<h3>7.5.3 Proportion of droplets (named by sample ID)</h3>
<pre class="r"><code># proportion of genetically assigned droplets in each HTO
p3 &lt;- ggcells(sce) + 
  geom_bar(
    aes(x = sampleID.HTO, fill = sampleID.genetics),
    position = position_fill(reverse = TRUE)) +
  coord_flip() +
  ylab(&quot;Frequency&quot;) +
  theme_cowplot(font_size = 8) + 
  scale_fill_manual(values = sce$colours$sampleID.genetics_colours)

# proportion of HTO assigned droplets in each genetic donor
p4 &lt;- ggcells(sce) + 
  geom_bar(
    aes(x = sampleID.genetics, fill = sampleID.HTO),
    position = position_fill(reverse = TRUE)) +
  coord_flip() +
  ylab(&quot;Frequency&quot;) +
  theme_cowplot(font_size = 8) + 
  scale_fill_manual(values = sce$colours$sampleID.HTO_colours)

((p3 + ggtitle(&quot;By HTO&quot;)) +
    p3 + facet_grid(~Capture) + plot_layout(widths = c(1, 2))) /
  ((p4 + ggtitle(&quot;By genetics&quot;)) + 
     p4 + facet_grid(~Capture) + plot_layout(widths = c(1, 2))) +
  plot_layout(guides = &quot;collect&quot;)</code></pre>
<p><img src="figure/02.qc_BP1.Rmd/unnamed-chunk-15-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
<div id="remove-undesired-genes" class="section level1">
<h1>8 Remove undesired genes</h1>
<pre class="r"><code># create uninformative gene sets
uninformative &lt;- is_mito | is_ribo | rownames(sce) %in% sex_set | rownames(sce) %in% pseudogene_set
sum(uninformative)</code></pre>
<pre><code>[1] 1608</code></pre>
<pre class="r"><code># remove uninformative genes 
sce &lt;- sce[!uninformative,]

# remove low-abundance genes 
numCells &lt;- nexprs(sce, byrow = TRUE)
keep &lt;- numCells &gt; 20
sum(keep)</code></pre>
<pre><code>[1] 23855</code></pre>
<pre class="r"><code>sce &lt;- sce[keep,]

# number of genes retained
dim(sce)</code></pre>
<pre><code>[1] 23855 40083</code></pre>
</div>
<div id="save-object" class="section level1">
<h1>9 Save object</h1>
<pre class="r"><code>prep_dir &lt;- here(&quot;data&quot;,&quot;SCEs&quot;,&quot;preprocessed&quot;)
if(!dir.exists(prep_dir)) {
  dir.create(prep_dir, recursive = TRUE)
}

out &lt;- paste0(prep_dir,&#39;/&#39;,
              paste0(batch_name,&quot;.preprocessed.SCE.rds&quot;))

if(!file.exists(out)) saveRDS(sce, out)</code></pre>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<p>Maksimovic J, Shanthikumar S, Howitt G, Hickey PF, Ho W, Anttila C,
et al.Â Single-cell atlas of bronchoalveolar lavage from preschool cystic
fibrosis reveals new cell phenotypes. bioRxiv 2022:
2022.2006.2017.496207.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.2 (2021-11-01)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS:   /hpc/software/installed/R/4.1.2/lib64/R/lib/libRblas.so
LAPACK: /hpc/software/installed/R/4.1.2/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] Homo.sapiens_1.3.1                     
 [2] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2
 [3] org.Hs.eg.db_3.14.0                    
 [4] GO.db_3.14.0                           
 [5] OrganismDbi_1.36.0                     
 [6] msigdbr_7.5.1                          
 [7] ensembldb_2.18.4                       
 [8] AnnotationFilter_1.22.0                
 [9] GenomicFeatures_1.46.4                 
[10] AnnotationDbi_1.56.2                   
[11] AnnotationHub_3.2.2                    
[12] BiocFileCache_2.6.1                    
[13] dbplyr_2.1.1                           
[14] stringr_1.5.1                          
[15] janitor_2.1.0                          
[16] forcats_0.5.1                          
[17] dplyr_1.1.4                            
[18] scater_1.22.0                          
[19] scuttle_1.4.0                          
[20] patchwork_1.2.0                        
[21] cowplot_1.1.3                          
[22] SeuratObject_5.0.1                     
[23] Seurat_4.4.0                           
[24] ggplot2_3.4.0                          
[25] here_1.0.1                             
[26] DropletUtils_1.14.2                    
[27] SingleCellExperiment_1.16.0            
[28] SummarizedExperiment_1.24.0            
[29] Biobase_2.54.0                         
[30] GenomicRanges_1.46.1                   
[31] GenomeInfoDb_1.30.1                    
[32] IRanges_2.28.0                         
[33] S4Vectors_0.32.4                       
[34] BiocGenerics_0.40.0                    
[35] MatrixGenerics_1.6.0                   
[36] matrixStats_1.1.0                      

loaded via a namespace (and not attached):
  [1] rappdirs_0.3.3                rtracklayer_1.54.0           
  [3] scattermore_1.2               R.methodsS3_1.8.1            
  [5] tidyr_1.2.0                   bit64_4.0.5                  
  [7] knitr_1.46                    irlba_2.3.5.1                
  [9] DelayedArray_0.20.0           R.utils_2.12.2               
 [11] data.table_1.15.4             KEGGREST_1.34.0              
 [13] RCurl_1.98-1.9                generics_0.1.3               
 [15] ScaledMatrix_1.2.0            RSQLite_2.3.6                
 [17] RANN_2.6.1                    future_1.33.2                
 [19] bit_4.0.5                     spatstat.data_3.0-4          
 [21] xml2_1.3.3                    lubridate_1.8.0              
 [23] httpuv_1.6.15                 assertthat_0.2.1             
 [25] viridis_0.6.2                 xfun_0.43                    
 [27] hms_1.1.1                     jquerylib_0.1.4              
 [29] babelgene_22.9                evaluate_0.23                
 [31] promises_1.3.0                fansi_1.0.6                  
 [33] restfulr_0.0.15               progress_1.2.2               
 [35] igraph_2.0.3                  DBI_1.2.2                    
 [37] htmlwidgets_1.6.4             spatstat.geom_3.2-9          
 [39] purrr_1.0.1                   ellipsis_0.3.2               
 [41] crosstalk_1.2.1               biomaRt_2.50.3               
 [43] deldir_2.0-4                  sparseMatrixStats_1.6.0      
 [45] vctrs_0.6.5                   ROCR_1.0-11                  
 [47] abind_1.4-5                   cachem_1.0.8                 
 [49] withr_3.0.0                   progressr_0.14.0             
 [51] sctransform_0.4.1             GenomicAlignments_1.30.0     
 [53] prettyunits_1.1.1             xaringanExtra_0.7.0          
 [55] goftest_1.2-3                 cluster_2.1.2                
 [57] dotCall64_1.1-1               lazyeval_0.2.2               
 [59] crayon_1.5.2                  spatstat.explore_3.2-7       
 [61] labeling_0.4.3                edgeR_3.36.0                 
 [63] pkgconfig_2.0.3               ProtGenerics_1.30.0          
 [65] nlme_3.1-155                  vipor_0.4.5                  
 [67] rlang_1.1.3                   globals_0.16.3               
 [69] lifecycle_1.0.4               miniUI_0.1.1.1               
 [71] filelock_1.0.3                rsvd_1.0.5                   
 [73] rprojroot_2.0.4               polyclip_1.10-6              
 [75] lmtest_0.9-40                 graph_1.72.0                 
 [77] Matrix_1.6-5                  Rhdf5lib_1.16.0              
 [79] zoo_1.8-12                    beeswarm_0.4.0               
 [81] ggridges_0.5.6                png_0.1-8                    
 [83] viridisLite_0.4.2             rjson_0.2.21                 
 [85] bitops_1.0-7                  R.oo_1.24.0                  
 [87] KernSmooth_2.23-20            spam_2.10-0                  
 [89] rhdf5filters_1.6.0            Biostrings_2.62.0            
 [91] blob_1.2.2                    DelayedMatrixStats_1.16.0    
 [93] workflowr_1.7.0               parallelly_1.37.1            
 [95] spatstat.random_3.2-3         beachmat_2.10.0              
 [97] scales_1.3.0                  memoise_2.0.1                
 [99] magrittr_2.0.3                plyr_1.8.9                   
[101] ica_1.0-3                     zlibbioc_1.40.0              
[103] compiler_4.1.2                dqrng_0.3.2                  
[105] BiocIO_1.8.0                  RColorBrewer_1.1-3           
[107] fitdistrplus_1.1-11           Rsamtools_2.10.0             
[109] snakecase_0.11.0              cli_3.6.2                    
[111] XVector_0.34.0                listenv_0.9.1                
[113] pbapply_1.7-2                 MASS_7.3-55                  
[115] tidyselect_1.2.1              stringi_1.8.3                
[117] highr_0.10                    yaml_2.3.8                   
[119] BiocSingular_1.10.0           locfit_1.5-9.4               
[121] ggrepel_0.9.1                 grid_4.1.2                   
[123] sass_0.4.9                    tools_4.1.2                  
[125] future.apply_1.11.2           parallel_4.1.2               
[127] rstudioapi_0.13               git2r_0.31.0                 
[129] gridExtra_2.3                 farver_2.1.1                 
[131] Rtsne_0.17                    digest_0.6.35                
[133] BiocManager_1.30.19           shiny_1.8.1.1                
[135] Rcpp_1.0.12                   BiocVersion_3.14.0           
[137] later_1.3.2                   RcppAnnoy_0.0.19             
[139] httr_1.4.7                    colorspace_2.1-0             
[141] XML_3.99-0.16.1               fs_1.6.4                     
[143] tensor_1.5                    reticulate_1.36.1.9000       
[145] splines_4.1.2                 RBGL_1.70.0                  
[147] uwot_0.1.14                   spatstat.utils_3.0-4         
[149] sp_2.1-3                      plotly_4.10.4.9000           
[151] xtable_1.8-4                  jsonlite_1.8.8               
[153] R6_2.5.1                      pillar_1.9.0                 
[155] htmltools_0.5.8.1             mime_0.12                    
[157] DT_0.20                       glue_1.7.0                   
[159] fastmap_1.1.1                 BiocParallel_1.28.3          
[161] BiocNeighbors_1.12.0          interactiveDisplayBase_1.32.0
[163] codetools_0.2-18              utf8_1.2.4                   
[165] lattice_0.20-45               bslib_0.3.1                  
[167] spatstat.sparse_3.0-3         tibble_3.2.1                 
[169] curl_5.2.1                    ggbeeswarm_0.6.0             
[171] leiden_0.4.3.1                survival_3.2-13              
[173] limma_3.50.3                  rmarkdown_2.26               
[175] munsell_0.5.1                 rhdf5_2.38.1                 
[177] GenomeInfoDbData_1.2.7        HDF5Array_1.22.1             
[179] reshape2_1.4.4                gtable_0.3.5                 </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
